### PR TITLE
Add an action to link sessions to customer ids.

### LINKED ISSUE #44852: [Enhancement]: New Action to fire before WC_Session_Handler converts a guest session to login.
### Describe the solution you'd like

I'd like to request an action be slipped in to fire either in in WC_Session_Handler::save_data() https://github.com/woocommerce/woocommerce/blob/487c7a87f0e31d85fd6cb1aaff45301cf7622a19/plugins/woocommerce/includes/class-wc-session-handler.php#L349-L375 or WC_Session_Handler::init_session_cookie() https://github.com/woocommerce/woocommerce/blob/487c7a87f0e31d85fd6cb1aaff45301cf7622a19/plugins/woocommerce/includes/class-wc-session-handler.php#L83-L123 that will let plugins listen for when a `guest_session_id` is getting blown away and its data transferred to a `user_id`.

Something like `do_action( 'wc_migrate_guest_session_key', $old_session_key, $user_id );` or the like.

This would let third-party code either connect the prior session key to a user id for data storage, or migrate the old session id to a different cookie that would enable discrete sessions even if multiple browsers authenticated to the same user_id.

### Describe alternatives you've considered

If this isn't considered, I've evaluated doing a much messier workaround, that being duplicating the session key in a side channel, and watching for it to change or for there be a logged in user concurrent with it, and try to ensure I'm catching the event before WooCommerce clears it, but it feels very ... klunky, liable to race conditions, and adding overhead to every page load.

### Additional context

For plugins that are providing additional context on user flows, the ability to be connect the pre-login result of `WC()->session->get_customer_unique_id()` to the post-login result where it returns a `user_id` instead can ensure a more seamless experience for customers without overcomplicated workarounds.

### PR DESCRIPTION
**Pull request by georgestephanis**
I'm unsure if the `@since` should be populated now, or if that happens when something merges -- I left it as `x.x.x` for now.

Unsure if this should move above the `::save_data()` invocation so it can fire before the migration happens -- but to have less potential for impacts, I'm adding it after the work is done so folks could have a record of what the id changed from and to.

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

 Fires an action after a customer has logged in, and their guest session id has been deleted with its data migrated to a customer id.

This hook gives extensions the chance to connect the old session id to the customer id, if the key is being used externally.

Closes #44852

### Testing

See https://github.com/woocommerce/woocommerce/pull/45146#issuecomment-1988772799.

### CODE REVIEW

**Comment by coreymckrill** [comment]
@georgestephanis could you add some testing instructions, including perhaps a code snippet that shows how the hook would be used?

**Comment by georgestephanis** [comment]
Sure.

Testing instructions:

Drop this code into a mu-plugins file:

```php
<?php

add_action( 'woocommerce_guest_session_to_user_id', function( $guest_session_id, $customer_id ) {
	wc_get_logger()->log(
		'info',
		sprintf( 'Guest with Session ID %s has logged in to Customer ID %s.', $guest_session_id, $customer_id ),
		array(
			'source' => 'test-pr-45146'
		)
	);

	// Potentially trigger other db writes or external api calls to any data stores that had been keying off the old session id.
}, 10, 2 );
```

Open a new incognito window and browse to your shop.  Add an item to the cart, and then go to log in.

After you log in, you should be able to view a record of that conversion from Session ID to Customer ID here: `/wp-admin/admin.php?page=wc-status&tab=logs`

Screenshot 2024-03-11 at 11 50 19 AM

Screenshot 2024-03-11 at 11 50 42 AM

File: plugins/woocommerce/includes/class-wc-session-handler.php [hunk 1]
```diff
@@ -108,6 +108,20 @@ public function init_session_cookie() {
...
 				$this->set_customer_session_cookie( true );
+
+				/**
+				 * Fires after a customer has logged in, and their guest session id has been
+				 * deleted with its data migrated to a customer id.
+				 *
+				 * This hook gives extensions the chance to connect the old session id to the
+				 * customer id, if the key is being used externally.
+				 *
+				 * @since x.x.x
```
Comment (coreymckrill) [inline plugins/woocommerce/includes/class-wc-session-handler.php]:
```suggestion
				 * @since 8.8.0
```

**Review by coreymckrill — APPROVED** [review]
Looks good, thank you for the testing instructions and code snippet!

> Unsure if this should move above the ::save_data() invocation so it can fire before the migration happens -- but to have less potential for impacts, I'm adding it after the work is done so folks could have a record of what the id changed from and to.

I think this is a good call. This way it's just a notification of something that happened instead of a place where the normal flow could get interrupted somehow.

**Pull request merged**